File: server/routes/share.js
const express = require('express');
const router = express.Router();
const Resume = require('../model/Resume');
const auth = require('../middlewares/auth');
const { generateShareToken } = require('../utils/share');
const mongoose = require('mongoose');

// Public: get resume by share token
router.get('/shares/:token', async (req, res) => {
  try {
    const { token } = req.params;
    const doc = await Resume.findOne({ 'share.token': token, 'share.enabled': true })
      .select({ name: 1, basics: 1, sections: 1, styles: 1, visibility: 1, 'share.expiresAt': 1 })
      .lean();

    if (!doc) return res.sendStatus(404);
    if (doc.share?.expiresAt && new Date() > new Date(doc.share.expiresAt)) {
      return res.sendStatus(404);
    }
    if (doc.visibility === 'private') return res.sendStatus(404);

    return res.json({ resume: doc });
  } catch (err) {
    return res.status(500).json({ message: err.message });
  }
});

// Change visibility: private | public | link
router.patch('/resumes/:id/visibility', auth, async (req, res) => {
  try {
    if (!req.auth || req.auth.type !== 'user') {
      return res.status(401).json({ message: 'Login required' });
    }
    const { id } = req.params;
    
    const { visibility } = req.body || {};
    console.log(id, visibility)
    console.log(new mongoose.Types.ObjectId(id), new mongoose.Types.ObjectId(req.auth.userId))
    if (!['private', 'public', 'link'].includes(visibility)) {
      return res.status(400).json({ message: 'Invalid visibility' });
    }

    const doc = await Resume.findOne({
      _id: new mongoose.Types.ObjectId(id),
      userId: new mongoose.Types.ObjectId(req.auth.userId)
    });

    if (!doc) return res.sendStatus(404);

    doc.visibility = visibility;
    if (visibility === 'private') {
      doc.share = { ...(doc.share || {}), enabled: false };
    }
    await doc.save();

    res.json({ ok: true, visibility: doc.visibility, shareEnabled: !!doc.share?.enabled });
  } catch (err) {
    return res.status(500).json({ message: err.message });
  }
});

// Enable share (create token if missing)
router.post('/resumes/:id/share/enable', auth, async (req, res) => {
  try {
    if (!req.auth || req.auth.type !== 'user') {
      return res.status(401).json({ message: 'Login required' });
    }
    const { id } = req.params;
    const doc = await Resume.findOne({ _id: id, userId: req.auth.userId });
    if (!doc) return res.sendStatus(404);

    const now = new Date();
    if (!doc.share?.token) {
      doc.share = { ...(doc.share || {}), token: generateShareToken(18) };
    }

    doc.share.enabled = true;
    doc.share.expiresAt = new Date(now.getTime() + 24 * 60 * 60 * 1000);
    doc.share.lastRotatedAt = now;
    await doc.save();

    res.json({ ok: true, token: doc.share.token, expiresAt: doc.share.expiresAt });
  } catch (err) {
    return res.status(500).json({ message: err.message });
  }
});

// Rotate token
router.post('/resumes/:id/share/rotate', auth, async (req, res) => {
  try {
    if (!req.auth || req.auth.type !== 'user') {
      return res.status(401).json({ message: 'Login required' });
    }
    const { id } = req.params;
    const doc = await Resume.findOne({ _id: id, userId: req.auth.userId });
    if (!doc) return res.sendStatus(404);

    const now = new Date();
    doc.share = { ...(doc.share || {}) };
    doc.share.token = generateShareToken(18);
    doc.share.enabled = true;
    doc.share.lastRotatedAt = now;
    doc.share.expiresAt = new Date(now.getTime() + 24 * 60 * 60 * 1000);
    await doc.save();

    res.json({ ok: true, token: doc.share.token, expiresAt: doc.share.expiresAt });
  } catch (err) {
    return res.status(500).json({ message: err.message });
  }
});

module.exports = router;

File: server_2/routes/share.py
from datetime import datetime, timedelta
from typing import Dict

from fastapi import APIRouter, Body, Depends, HTTPException, status
from bson import ObjectId

from controllers import resume_controller
from database import get_database
from middlewares.auth import AuthContext, require_auth
from utils import generate_share_token

router = APIRouter()


@router.get("/shares/{token}")
async def get_shared_resume(token: str, db=Depends(get_database)):
    doc = await db.resumes.find_one(
        {"share.token": token, "share.enabled": True},
        projection={"name": 1, "basics": 1, "sections": 1, "styles": 1, "visibility": 1, "share.expiresAt": 1},
    )
    if not doc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Not found")

    expires_at = doc.get("share", {}).get("expiresAt")
    if expires_at and datetime.utcnow() > expires_at:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Not found")
    if doc.get("visibility") == "private":
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Not found")

    return {"resume": resume_controller.serialize_resume(doc)}


@router.patch("/resumes/{id}/visibility")
async def change_visibility(
    id: str,
    payload: Dict = Body(...),
    auth: AuthContext = Depends(require_auth),
    db=Depends(get_database),
):
    if auth.get("type") != "user" or not auth.get("userId"):
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Login required")

    visibility = payload.get("visibility")
    if visibility not in {"private", "public", "link"}:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid visibility")

    try:
        oid = ObjectId(id)
        owner_id = ObjectId(auth["userId"])
    except Exception as exc:  # pylint: disable=broad-except
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid resume id") from exc
    doc = await db.resumes.find_one({"_id": oid, "userId": owner_id})
    if not doc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Not found")

    updates = {"visibility": visibility, "updatedAt": datetime.utcnow()}
    if visibility == "private":
        updates["share.enabled"] = False
    await db.resumes.update_one({"_id": oid, "userId": owner_id}, {"$set": updates})

    return {"ok": True, "visibility": visibility, "shareEnabled": visibility != "private"}


@router.post("/resumes/{id}/share/enable")
async def enable_share(
    id: str,
    auth: AuthContext = Depends(require_auth),
    db=Depends(get_database),
):
    if auth.get("type") != "user" or not auth.get("userId"):
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Login required")

    try:
        oid = ObjectId(id)
        owner_id = ObjectId(auth["userId"])
    except Exception as exc:  # pylint: disable=broad-except
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid resume id") from exc
    doc = await db.resumes.find_one({"_id": oid, "userId": owner_id})
    if not doc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Not found")

    now = datetime.utcnow()
    share = doc.get("share") or {}
    token = share.get("token") or generate_share_token(18)
    expires_at = now + timedelta(hours=24)
    await db.resumes.update_one(
        {"_id": oid, "userId": owner_id},
        {
            "$set": {
                "share.token": token,
                "share.enabled": True,
                "share.expiresAt": expires_at,
                "share.lastRotatedAt": now,
                "updatedAt": now,
            }
        },
    )
    return {"ok": True, "token": token, "expiresAt": expires_at}


@router.post("/resumes/{id}/share/rotate")
async def rotate_share(
    id: str,
    auth: AuthContext = Depends(require_auth),
    db=Depends(get_database),
):
    if auth.get("type") != "user" or not auth.get("userId"):
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Login required")

    try:
        oid = ObjectId(id)
        owner_id = ObjectId(auth["userId"])
    except Exception as exc:  # pylint: disable=broad-except
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid resume id") from exc
    doc = await db.resumes.find_one({"_id": oid, "userId": owner_id})
    if not doc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Not found")

    now = datetime.utcnow()
    token = generate_share_token(18)
    expires_at = now + timedelta(hours=24)
    await db.resumes.update_one(
        {"_id": oid, "userId": owner_id},
        {
            "$set": {
                "share.token": token,
                "share.enabled": True,
                "share.expiresAt": expires_at,
                "share.lastRotatedAt": now,
                "updatedAt": now,
            }
        },
    )
    return {"ok": True, "token": token, "expiresAt": expires_at}
